File: python-solid-principle-sample-code/com/principle/solid/dependency_inversion/ShapePrinter.py
from IAreaCalculator import IAreaCalculator


class ShapePrinter:

    """
        Dependency Inversion Violation - Taking Concreted Class Reference.
    """
    # def __init__(self, area_calculator: AreaCalculator):
    #     self.__area_calculator = area_calculator

    """
        Dependency Inversion Achieved - Instead of Concreted Class take a Interface Reference and Inject on-demand.
    """
    def __init__(self, area_calculator: IAreaCalculator):
        self.__area_calculator = area_calculator

    def json(self, shapes):
        return "{{shapes_sum: {}}}".format(int(self.__area_calculator.sum(shapes)))

    def csv(self, shapes):
        return "shapes_sum,{}".format(int(self.__area_calculator.sum(shapes)))

File: java-solid-principle-sample-code/src/com/principle/solid/dependency_inversion/ShapePrinter.java
package com.principle.solid.dependency_inversion;

import java.util.List;

public class ShapePrinter {

	/**
	 	Dependency Inversion Violation - Taking Concreted Class Reference.
	
	private AreaCalculator areaCalculator = null;
	public ShapePrinter(AreaCalculator areaCalculator) {
        this.areaCalculator = areaCalculator;
    }
	*/
	
	/**
	 	Dependency Inversion Achieved - Instead of Concreted Class take a Interface Reference and Inject on-demand.
	 */
    private IAreaCalculator areaCalculator = null;

    public ShapePrinter(IAreaCalculator areaCalculator) {
        this.areaCalculator = areaCalculator;
    }

    public String json(List<Shape> shapes) {
        return String.format("{shapes_sum: %s}", areaCalculator.sum(shapes));
    }

    public String csv(List<Shape> shapes) {
        return String.format("shapes_sum,%s", areaCalculator.sum(shapes));
    }
}
